src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractParallelFixAllProvider.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeStyle;
+
+namespace Microsoft.CodeAnalysis.Shared.CodeFixes
+{
+    /// <summary>
+    /// A simple implementation of a <see cref="FixAllProvider"/> that handles parallelizing
+    /// execution over all the <see cref="Document"/>s in a <see cref="Project"/>.
+    /// <para/>
+    /// Subclasses need only concern them selves with processing all the diagnostics for a specific
+    /// <see cref="Document"/>.  This class will then merge all the disparate changes to all
+    /// those documents into a full solution change.
+    /// </summary>
+    internal abstract class AbstractParallelFixAllProvider : FixAllProvider
```

Comment:
this was helpful code extracted from SyntaxEditorBasedFixAllProvider.  The latter is for when you spefically would like to work with a common Syntax-Editor across all diagnostics.  This type is if you don't need htat functionality and just want the benefits of parallel processing of documents.

Reply:
You should be able to use the existing type `DocumentBasedFixAllProvider` for this.

Reply:
hrmm.  we should try to reconcile these.  There are definitely places they've diverged.  For example, these helpers sort and filter the diagnostics nicely.  let me see what i can do here.